File: ruby-app/version1_soap.rb
require 'sinatra'
require 'savon'

get '/soap' do
  numero = params[:n] || "10"
  
  begin
    client = Savon.client(wsdl: "https://www.dataaccess.com/webservicesserver/NumberConversion.wso?WSDL")
    respuesta = client.call(:number_to_words, message: { ubiNum: numero })
    resultado = respuesta.body[:number_to_words_response][:number_to_words_result]
    
    resultado
    
  rescue => e
    # Fallback en caso de timeout del servicio externo
    mock = { "1" => "one", "2" => "two", "10" => "ten", "20" => "twenty" }
    mock[numero] || "number"
  end
end

puts "Servidor Ruby en http://localhost:4567"
puts "Prueba: http://localhost:4567/soap?n=10"

File: perl-app/version1_soap.pl
#!/usr/bin/perl
use strict;
use warnings;
use SOAP::Lite;
use HTTP::Daemon;

my $client = SOAP::Lite
    -> uri('http://www.dataaccess.com/webservicesserver/')
    -> proxy('https://www.dataaccess.com/webservicesserver/NumberConversion.wso');

my $server = HTTP::Daemon->new(
    LocalAddr => 'localhost',
    LocalPort => 5000,
) or die "No se pudo iniciar el servidor: $!";

print "Servidor Perl SOAP en http://localhost:5000\n";

while (my $conn = $server->accept) {
    my $request = $conn->get_request;
    my $uri = $request->uri;
    
    my $numero = 10;
    $numero = $1 if $uri =~ /n=(\d+)/;
    
    my $resultado = "number";
    eval {
        my $respuesta = $client->call(NumberToWords => SOAP::Data->name(ubiNum => $numero));
        $resultado = $respuesta->result if $respuesta->result;
    };
    
    # Fallback por si el servicio externo presenta latencia
    my %mock = (1 => "one", 2 => "two", 10 => "ten", 20 => "twenty");
    $resultado = $mock{$numero} if $resultado eq "number" || !$resultado;
    
    $conn->send_response(HTTP::Response->new(200, 'OK', ['Content-Type' => 'text/plain'], $resultado));
    $conn->close;
}